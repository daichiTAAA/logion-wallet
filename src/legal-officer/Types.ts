import { ColorTheme, rgbaToHex } from '../component/ColorTheme';
import Identity from '../component/types/Identity';
import PostalAddress from '../component/types/PostalAddress';

export const DEFAULT_LEGAL_OFFICER = "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"; // Alice
export const ANOTHER_LEGAL_OFFICER = "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty"; // Bob

const LEGAL_OFFICERS: string[] = [ DEFAULT_LEGAL_OFFICER, ANOTHER_LEGAL_OFFICER ];

export function isLegalOfficer(address: string): boolean {
    return LEGAL_OFFICERS.includes(address);
}

export type TokenizationRequestStatus = "PENDING" | "REJECTED" | "ACCEPTED";

export interface AssetDescription {
    assetId: string,
    decimals: number,
}

export interface TokenizationRequest {
    id: string,
    legalOfficerAddress: string,
    requesterAddress: string,
    requestedTokenName: string,
    bars: number,
    status: TokenizationRequestStatus,
    rejectReason?: string | null,
    createdOn?: string,
    decisionOn?: string,
    assetDescription?: AssetDescription,
}

export type LegalOfficerDecisionStatus = "PENDING" | "REJECTED" | "ACCEPTED";

export interface LegalOfficerDecision {
    legalOfficerAddress: string,
    status: LegalOfficerDecisionStatus,
    rejectReason: string | null,
    decisionOn: string | null,
}

export interface ProtectionRequest {
    id: string,
    requesterAddress: string,
    decisions: LegalOfficerDecision[],
    userIdentity: Identity,
    userPostalAddress: PostalAddress,
    createdOn: string
}

export const LIGHT_MODE: ColorTheme = {
    type: 'light',
    shadowColor: rgbaToHex('#3b6cf4', 0.1),
    dashboard: {
        background: rgbaToHex('#3b6cf4', 0.05),
        foreground: '#000000',
    },
    sidebar: {
        background: '#ffffff',
        foreground: '#000000',
        activeItemBackground: rgbaToHex('#3b6cf4', 0.15),
    },
    accounts: {
        iconBackground: '#3b6cf4',
        hintColor: rgbaToHex('#000000', 0.6),
        foreground: '#000000',
        background: rgbaToHex('#3b6cf4', 0.05),
        legalOfficerIcon: {
            category: 'legal-officer',
            id: 'account-shield'
        }
    },
    frame: {
        background: '#ffffff',
        foreground: '#000000',
        link: rgbaToHex('#3b6cf4', 0.15),
    },
    topMenuItems: {
        iconGradient: {
            from: '#3b6cf4',
            to: '#6050dc',
        }
    },
    bottomMenuItems: {
        iconGradient: {
            from: '#7a90cb',
            to: '#3b6cf4',
        }
    }
};
