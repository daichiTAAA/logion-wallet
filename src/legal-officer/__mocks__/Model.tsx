import {
    fetchRecoveryInfo,
} from './ModelMock';

export {
    fetchRecoveryInfo,
};
