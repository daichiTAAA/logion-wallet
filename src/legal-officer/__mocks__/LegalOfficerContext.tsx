export let pendingTokenizationRequests: any[] | null = null;

export let rejectRequest = () => {

};

export let acceptedTokenizationRequests: any[] | null = null;

export let rejectedTokenizationRequests: any[] | null = null;

export let refreshRequests = () => {

};

export let pendingProtectionRequests: any[] | null = null;

export let acceptedProtectionRequests: any[] | null = null;

export let protectionRequestsHistory: any[] | null = null;

export function useLegalOfficerContext() {
    return {
        pendingTokenizationRequests,
        rejectRequest,
        rejectedTokenizationRequests,
        acceptedTokenizationRequests,
        refreshRequests,
        pendingProtectionRequests,
        acceptedProtectionRequests,
        protectionRequestsHistory,
    };
}

export function setPendingRequests(requests: any[]) {
    pendingTokenizationRequests = requests;
}

export function setRejectRequest(callback: any) {
    rejectRequest = callback;
}

export function setRejectedRequests(requests: any[]) {
    rejectedTokenizationRequests = requests;
}

export function setAcceptedRequests(requests: any[]) {
    acceptedTokenizationRequests = requests;
}

export function setRefreshRequests(callback: any) {
    refreshRequests = callback;
}

export function setPendingProtectionRequests(requests: any[]) {
    pendingProtectionRequests = requests;
}

export function setAcceptedProtectionRequests(requests: any[]) {
    acceptedProtectionRequests = requests;
}

export function setProtectionRequestsHistory(requests: any[]) {
    protectionRequestsHistory = requests;
}
