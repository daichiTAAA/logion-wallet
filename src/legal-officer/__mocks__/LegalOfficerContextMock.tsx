export let pendingTokenizationRequests: any[] | null = null;

export let rejectRequest = () => {

};

export let acceptedTokenizationRequests: any[] | null = null;

export let rejectedTokenizationRequests: any[] | null = null;

export let refreshRequests = jest.fn();

export let pendingProtectionRequests: any[] | null = null;

export let activatedProtectionRequests: any[] | null = null;

export let protectionRequestsHistory: any[] | null = null;

export let recoveryRequestsHistory: any[] | null = null;

export function setPendingRequests(requests: any[]) {
    pendingTokenizationRequests = requests;
}

export function setRejectRequest(callback: any) {
    rejectRequest = callback;
}

export function setRejectedRequests(requests: any[]) {
    rejectedTokenizationRequests = requests;
}

export function setAcceptedRequests(requests: any[]) {
    acceptedTokenizationRequests = requests;
}

export function setRefreshRequests(callback: any) {
    refreshRequests = callback;
}

export function setPendingProtectionRequests(requests: any[]) {
    pendingProtectionRequests = requests;
}

export function setActivatedProtectionRequests(requests: any[]) {
    activatedProtectionRequests = requests;
}

export function setProtectionRequestsHistory(requests: any[]) {
    protectionRequestsHistory = requests;
}

export function setRecoveryRequestsHistory(requests: any[]) {
    recoveryRequestsHistory = requests;
}
