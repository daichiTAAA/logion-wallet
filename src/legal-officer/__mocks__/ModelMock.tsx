export const DEFAULT_LEGAL_OFFICER = "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"; // Alice
export const ANOTHER_LEGAL_OFFICER = "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty"; // Bob

export let rejectRequest = jest.fn();

export let acceptRequest = jest.fn();

export function setAcceptRequest(mockFn: any) {
    acceptRequest = mockFn;
}

export let setAssetDescription = jest.fn();

export let rejectProtectionRequest = jest.fn();

export function setRejectProtectionRequest(mockFn: any) {
    rejectProtectionRequest = mockFn;
}

export let acceptProtectionRequest = jest.fn();

export function setAcceptProtectionRequest(mockFn: any) {
    acceptProtectionRequest = mockFn;
}

export function decision(legalOfficerAddress: string, decisions: any[]) {
    return decisions[0];
}

export let fetchRecoveryInfo = jest.fn();

export function setFetchRecoveryInfo(mockFn: any) {
    fetchRecoveryInfo = mockFn;
}
