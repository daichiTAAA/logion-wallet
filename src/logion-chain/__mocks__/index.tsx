import { Extension } from '../Keys';

const LogionChainContextProvider = (props: any) => null;

let context = {
    apiState: 'CONNECT_INIT',
    injectedAccounts: []
};

export function setContextMock(value: any) {
    context = value;
}

export function useLogionChain() {
    return context;
}

let extensionAvailable = false;

export function setExtensionAvailable(value: boolean) {
    extensionAvailable = value;
}

export function isExtensionAvailable() {
    return extensionAvailable;
}

let recommendedExtensionValue: Extension | null = null;

export function setRecommendedExtension(value: Extension | null) {
    recommendedExtensionValue = value;
}

export function recommendedExtension(): Extension | null {
    return recommendedExtensionValue;
}

export {
    LogionChainContextProvider,
};
