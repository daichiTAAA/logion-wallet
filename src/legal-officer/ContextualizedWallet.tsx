import React, { useCallback } from 'react';

import Dashboard from '../common/Dashboard';

import LegalOfficerRouter from './LegalOfficerRouter';
import { useCommonContext } from '../common/CommonContext';

import {
    HOME_PATH,
    PROTECTION_REQUESTS_PATH,
    RECOVERY_REQUESTS_PATH,
    SETTINGS_PATH,
    WALLET_PATH,
    IDENTITIES_PATH, locRequestsPath,
} from './LegalOfficerPaths';
import { useLegalOfficerContext } from './LegalOfficerContext';

export default function ContextualizedWallet() {
    const { selectAddress, accounts, colorTheme, refresh, availableLegalOfficers } = useCommonContext();
    const { refreshRequests } = useLegalOfficerContext();

    const refreshAll = useCallback(() => {
        refresh(false);
        refreshRequests!(false);
    }, [ refresh, refreshRequests ]);

    if(accounts === null || selectAddress === null || availableLegalOfficers === undefined) {
        return null;
    }

    const currentLegalOfficerUnavailable = availableLegalOfficers.find(node => node.address === accounts?.current?.address) === undefined;

    return (
        <Dashboard
            menuTop={[
                {
                    id: "home",
                    text: "Home",
                    to: HOME_PATH,
                    exact: true,
                    icon: {
                        icon: {
                            id: 'home'
                        },
                        background: colorTheme.topMenuItems.iconGradient,
                    },
                    onClick: refreshAll,
                    disabled: currentLegalOfficerUnavailable,
                },
                {
                    id: "wallet",
                    text: "Wallet",
                    to: WALLET_PATH,
                    exact: false,
                    icon: {
                        icon: {
                            id: 'wallet'
                        },
                        background: colorTheme.topMenuItems.iconGradient,
                    },
                    onClick: refreshAll,
                    disabled: currentLegalOfficerUnavailable,
                },
                {
                    id: "loc-collection",
                    text: "Collections",
                    to: locRequestsPath('Collection'),
                    exact: false,
                    icon: {
                        icon: {
                            id: 'collection'
                        },
                        background: colorTheme.topMenuItems.iconGradient,
                    },
                    onClick: refreshAll,
                    disabled: currentLegalOfficerUnavailable,
                },
                {
                    id: "loc-transaction",
                    text: "Transactions",
                    to: locRequestsPath('Transaction'),
                    exact: false,
                    icon: {
                        icon: {
                            id: 'loc'
                        },
                        background: colorTheme.topMenuItems.iconGradient,
                    },
                    onClick: refreshAll,
                    disabled: currentLegalOfficerUnavailable,
                },
                {
                    id: "identity",
                    text: "Identities",
                    to: IDENTITIES_PATH,
                    exact: false,
                    icon: {
                        icon: {
                            id: 'identity'
                        },
                        background: colorTheme.topMenuItems.iconGradient,
                    },
                    onClick: refreshAll,
                    disabled: currentLegalOfficerUnavailable,
                }
            ]}
            menuMiddle={[
                {
                    id: "protection",
                    text: "Protection Management",
                    to: PROTECTION_REQUESTS_PATH,
                    exact: true,
                    icon: {
                        icon: {
                            id: 'shield',
                            hasVariants: true,
                        },
                        height: 'auto',
                        width: '60px',
                    },
                    onClick: refreshAll,
                    disabled: currentLegalOfficerUnavailable,
                },
                {
                    id: "recovery",
                    text: "Recovery Requests",
                    to: RECOVERY_REQUESTS_PATH,
                    exact: true,
                    icon: {
                        icon: {
                            id: 'recovery_request',
                            hasVariants: true,
                        },
                        height: 'auto',
                        width: '60px',
                    },
                    onClick: refreshAll,
                    disabled: currentLegalOfficerUnavailable,
                }
            ]}
            menuBottom={[
                {
                    id: "settings",
                    text: "Settings",
                    to: SETTINGS_PATH,
                    exact: true,
                    icon: {
                        icon: {
                            id: 'settings'
                        },
                        background: colorTheme.bottomMenuItems.iconGradient,
                    },
                    onClick: refreshAll,
                }
            ]}
        >
            <LegalOfficerRouter />
        </Dashboard>
    );
}
