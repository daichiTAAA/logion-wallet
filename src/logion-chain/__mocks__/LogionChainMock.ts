import BN from 'bn.js'
import { UUID } from "../UUID";
import { mockCompact } from "../../__mocks__/PolkadotApiMock";

export const LogionChainContextProvider = (props: any) => null;

const DEFAULT_BALANCE = new BN("42000000000000000000");

const DEFAULT_RECOVERY_CONFIG = {
    isSome: true,
    unwrap: () => {}
}

const DEFAULT_ACTIVE_RECOVERY = {
    isSome: true,
    unwrap: () => {}
}

const DEFAULT_COLLECTION_ITEM = {
    isSome: true,
    unwrap: () => {
        return {
            description: {
                toUtf8: () => {
                    return "something"
                }
            }
        }
    }
}

const DEFAULT_COLLECTION_SIZE = {
    isSome: true,
    unwrap: () => {
        return mockCompact(1)
    }
}

const api = {
    tx: {
        assets: {
            create: () => {}
        },
        recovery: {
            asRecovered: () => {},
        }
    },
    query: {
        assets: {
            account: (assetId: any, account: any) => {
                return Promise.resolve({
                    balance: DEFAULT_BALANCE
                });
            }
        },
        recovery: {
            recoverable: (accountId: any) => {
                return Promise.resolve(DEFAULT_RECOVERY_CONFIG);
            },
            activeRecoveries: (source: string, dest: string) => {
                return Promise.resolve(DEFAULT_ACTIVE_RECOVERY);
            },
        },
        logionLoc: {
            collectionSizeMap: (locId: UUID) => {
                return Promise.resolve(DEFAULT_COLLECTION_SIZE)
            },
            collectionItemsMap: (locId: UUID, itemId: string) => {
                return Promise.resolve(DEFAULT_COLLECTION_ITEM)
            }
        }
    }
};

let context = {
    apiState: 'CONNECT_INIT',
    injectedAccounts: null,
    api,
    connectedNodeMetadata: {

    }
};

export function setContextMock(value: any) {
    context = value;
}

export function useLogionChain() {
    return context;
}
