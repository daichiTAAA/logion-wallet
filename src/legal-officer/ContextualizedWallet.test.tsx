jest.mock("../logion-chain");
jest.mock("../RootContext");

import React from 'react';
import ContextualizedWallet from './ContextualizedWallet';
import { DEFAULT_LEGAL_OFFICER } from './Types';
import { shallowRender, mockAccount } from '../tests';
import { setContextMock } from '../logion-chain/__mocks__/LogionChainMock';

test('renders', () => {
    setContextMock({
        injectedAccounts: [
            mockAccount(DEFAULT_LEGAL_OFFICER, "Account name")
        ]
    });
    const tree = shallowRender(<ContextualizedWallet />);
    expect(tree).toMatchSnapshot();
});
