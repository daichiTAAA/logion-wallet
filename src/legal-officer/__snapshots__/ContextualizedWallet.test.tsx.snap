// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<Dashboard
  colors={
    Object {
      "accounts": Object {
        "background": "#141b2d",
        "foreground": "#ffffff",
        "hintColor": "#ffffff66",
        "iconBackground": "#3b6cf4",
        "legalOfficerIcon": Object {
          "category": "legal-officer",
          "id": "account-shield",
        },
      },
      "bottomMenu": Object {
        "iconGradient": Object {
          "from": "#7a90cb",
          "to": "#3b6cf4",
        },
      },
      "dashboard": Object {
        "background": "#141b2d",
        "foreground": "#ffffff",
      },
      "frame": Object {
        "background": "#1f2a40",
        "foreground": "#ffffff",
      },
      "menuArea": Object {
        "background": "#3b6cf40f",
        "foreground": "#ffffff",
      },
      "primaryArea": Object {
        "background": "#141b2d",
        "foreground": "#ffffff",
        "link": "#ffffff",
      },
      "secondaryArea": Object {
        "background": "#141b2d",
        "foreground": "#ffffff",
      },
      "shadowColor": "#3b6cf419",
      "topMenu": Object {
        "iconGradient": Object {
          "from": "#3b6cf4",
          "to": "#6050dc",
        },
      },
      "type": "dark",
    }
  }
  menuBottom={
    Array [
      Object {
        "exact": true,
        "icon": Object {
          "icon": Object {
            "category": "legal-officer",
            "id": "settings",
          },
        },
        "text": "Settings",
        "to": "/legal-officer/settings",
      },
    ]
  }
  menuTop={
    Array [
      Object {
        "exact": true,
        "icon": Object {
          "icon": Object {
            "category": "legal-officer",
            "id": "tokens",
          },
        },
        "text": "Tokens",
        "to": "/legal-officer/tokenization",
      },
    ]
  }
  selectAddress={[MockFunction]}
  shieldItem={
    Object {
      "exact": true,
      "icon": Object {
        "height": "auto",
        "icon": Object {
          "category": "legal-officer",
          "id": "shield",
        },
        "width": "auto",
      },
      "text": "Protection Requests",
      "to": "/legal-officer/protection",
    }
  }
>
  <LegalOfficerRouter />
</Dashboard>
`;
